src/analysis/processing/qgsalgorithmspatialjoin.cpp

```diff
@@ -0,0 +1,375 @@
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "UNJOINED" ), QObject::tr( "unjoined layer" ), QgsProcessing::TypeVectorAnyGeometry, QVariant(), true, true ) );
+}
+
+QString QgsSpatialJoinAlgorithm::name() const
+{
+  return QStringLiteral( "spatialjoin" );
+}
+
+QString QgsSpatialJoinAlgorithm::displayName() const
+{
+  return QObject::tr( "Join attributes by their spatial relation" );
+}
+
+QStringList QgsSpatialJoinAlgorithm::tags() const
+{
+  return QObject::tr( "intersect,contain,isEqual,touch,overlap,within,cross" ).split( ',' );
+}
+
+QString QgsSpatialJoinAlgorithm::svgIconPath() const
+{
+  return QgsApplication::iconPath( QStringLiteral( "/algorithms/... .svg" ) );
+}
+
+QIcon QgsSpatialJoinAlgorithm::icon() const
+{
+  return QgsApplication::getThemeIcon( QStringLiteral( "/algorithms/... .svg" ) );
+}
+
+QString QgsSpatialJoinAlgorithm::group() const
+{
+  return QObject::tr( "Vector analysis" );
+}
+
+QString QgsSpatialJoinAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectoranalysis" );
+}
+
+QString QgsSpatialJoinAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm takes an input vector layer and creates a new vector layer "
+                      "that is an extended version of the input one, with additional attributes in its attribute table.\n\n"
+                      "The additional attributes and their values are taken from a second vector layer. "
+                      "A spatial criteria is applied to select the values from the second layer that are added "
+                      "to each feature from the first layer in the resulting one." );
+}
+
+QString QgsSpatialJoinAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Join attributes from one vector layer to another by location." );
+}
+
+QgsSpatialJoinAlgorithm *QgsSpatialJoinAlgorithm::createInstance() const
+{
+  return new QgsSpatialJoinAlgorithm();
+}
+
+QList<int> QgsSpatialJoinAlgorithm::inputLayerTypes() const
+{
+  return QList< int >() << QgsProcessing::TypeVectorAnyGeometry;
+}
+
+QgsProcessing::SourceType QgsSpatialJoinAlgorithm::outputLayerType() const
+{
+  return QgsProcessing::TypeVectorAnyGeometry;
+}
+
+QgsCoordinateReferenceSystem QgsSpatialJoinAlgorithm::outputCrs( const QgsCoordinateReferenceSystem &inputCrs ) const
+{
+  mCrs = inputCrs;
+  return mCrs;
+}
+
+QString QgsSpatialJoinAlgorithm::inputParameterName() const
+{
+  return QStringLiteral( "Base layer" );
+}
+
+QString QgsSpatialJoinAlgorithm::inputParameterDescription() const
+{
+  return QObject::tr( "Base layer to join attribute to." );
+}
+
+QString QgsSpatialJoinAlgorithm::outputName() const
+{
+  return QObject::tr( "Joined layer" );
+}
+
+bool QgsSpatialJoinAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  std::unique_ptr<QgsProcessingFeatureSource> mBaseSource( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
```

Comment:
The definition of mBaseSource and mJoinSource needs to be in the header, not here (otherwise they'll go out of scope as soon as this method exits) 

(did this even build?)

Reply:
Yes it did build but I edited the code in two distinct environment and that made some weird partial changes to the code.

Reply:
They are declared in the header also.

Reply:
Right, but by re-declaring them here, you're using the local variables only -- as soon as this function exits they'll be deleted.

Reply:
I forgot that I redeclared many things and forgot to fix them, this lead to many issues, should be fixed now though. thanks for reminding me!